### PR TITLE
Make sure consistent id used to defer _getRead and _getWrite

### PR DESCRIPTION
**Pull request by rob-c**
Make sure the ```_inprogressDict``` keeps a consistent set of ```id()``` to allow for ```_getReadAccess``` ```_getWriteAccess``` calls at the repo level from causing problems. Possibly alleviates #374.

This was quicker to fix/demonstrate with the PR, if anyone has any further questions please ask.

### CODE REVIEW

**Comment by milliams** [comment]
So do I understand that with this change, `self.changed_ids` has the registry id of an object as its key but `self._inprogressDict` has the python `id()`?

I'm not sure why `find()` is not just:
```python
def find(self, _obj):
    return next(for obj in self._objects if obj is _obj)
```
(with maybe a smidgeon more error handling)?

**Comment by rob-c** [comment]
a) because the find method will fail when adding the object to the registry. It also makes no sense to get a read/write lock on an object as you're adding it to the dictionary, as it makes no sense to get access while flushing or something similar.

b) aside from the error handling and assert statements this is what the find is? (I think)
The problem with doing literally what you suggest is that this becomes extremely expensive and requires correctly comparing the full content of a job or a partly-loaded job. Working out how to do this without violating lazy-loading is potentially very difficult imo.
If we get the object equivalency sorted out then I think we can make the find method simpler.

**Comment by milliams** [comment]
a) Where are we calling `find` before adding a job to the registry? I agree that we shouldn't do that.

b) My problem with the current `find()` method is the chunk that looks like:
```python
...
if hasattr(obj, _reg_id_str):
    obj_reg_id = getattr(obj, _reg_id_str)
    objects_obj = self._objects[obj_reg_id]
    assert obj == objects_obj
    if hasattr(obj, _id_str):
        if hasattr(objects_obj, _id_str):
            assert getattr(obj, _id_str) == getattr(objects_obj, _id_str)
    assert obj_reg_id == getattr(objects_obj, _reg_id_str)
    return obj_reg_id
elif hasattr(obj, _id_str):
    obj_id = getattr(obj, _id_str)
    objects_obj = self._objects[obj_id]
    if hasattr(objects_obj, _reg_id_str):
        assert obj_id == getattr(objects_obj, _reg_id_str)
    assert obj_id == getattr(objects_obj, _id_str)
    return obj_id
...
```
It's messy and repetitive and it's not clear what it's doing. I don't think we should be storing on the object what its registry id is as that is duplicating information which then needs to be kept in line.

There's no issue with the object equality as my example uses `is` rather than `==`. It's saying 'is this object in the dict, if so give us the `id`'

**Comment by rob-c** [comment]
And strictly __not__ the point of this PR.

If we want to clean this then OK, but all I'm saying is that we should be using the ```id()``` method with the dictionary rather than making the repo call as that returns the ```obj.id``` I'm making no statement about the find method which can be changed after this.

**Comment by rob-c** [comment]
a) the inprogress stuff stops a call to _getReadAccess or _getWriteAccess which just confuses things during the add process and at worst can trigger loading of a half-written job from disk as the XML is being defined. It should use ```id()``` and not ```self.find()```

**Comment by milliams** [comment]
Ok, I'm not arguing against this PR but simply thinking aloud about exactly why these things keep causing us problems.

I'm happy for this PR to go in once the declaration of `self._inprogressDict` in `__init__` has a comment making clear that the keys of the dictionary should be Python `id()`s rather than registry ids.

**Comment by rob-c** [comment]
OK, I can add the comment but I'm just wanting to avoid touching anything here other than fixing a specific problem.

File: python/Ganga/Core/GangaRepository/Registry.py [hunk 1]
```diff
@@ -252,7 +252,7 @@ def __init__(self, name, doc, dirty_flush_counter=10, update_index_time=30, dirt
         self._dirty_max_timeout = dirty_max_timeout
         self._dirty_min_timeout = dirty_min_timeout
 
-
+        ## Id's id(obj) of objects undergoing a transaction such as flush, remove, add, etc.
```
Comment (milliams) [inline python/Ganga/Core/GangaRepository/Registry.py]: PEP8 specifies a single `#` for line comments. No need to change it here but for future...

**Comment by milliams** [comment]
Once the tests pass :+1:

**Pull request merged**